### ISSUE TITLE
Error with type narrowing for TypeDict fields

### ISSUE THREAD

**Issue by Daggy1234** [microsoft/pyright#2174]
**Describe the bug**

Type narrowing doesn't occur for TypedDict fields

**To Reproduce**

```py
from typing import Optional, TypedDict


class Td(TypedDict):
	opt: Optional[str]


def test_func():
	typed_dict: Td = {
		"opt": "test"
	}

	if typed_dict["opt"] is None:
		return "no"
	return "yes" + typed_dict["opt"]

```

**Expected behavior**

optda["opt"] should narrow to a str

**Screenshots or Code**

<img width="914" alt="image" src="https://user-images.githubusercontent.com/60603110/129055567-a9bfbb4f-b867-489a-b328-757075f0ec8e.png">

**VS Code extension or command-line**

Sublimelsp with pyright. Can reproduce with vscode pylance.

**Comment by erictraut** [comment]
Thanks for reporting this. It's a side effect of [this recent change](https://github.com/microsoft/pyright/issues/2087) where we decided to be much more conservative about type narrowing for subscript expressions.

In the case of TypedDict, it is safe to perform narrowing, so I've updated the code to handle this case. This fix will be in the next release.

**Comment by Daggy1234** [comment]
Ah thanks! I'll close this issue then

**Comment by erictraut** [comment]
Thanks, but I'm going to leave it open until the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.162, which I just published. It will also be in the next release of pylance.
